### PR TITLE
changes based on pull request 915

### PR DESCRIPTION
**Pull request by heinonline**
changes based on pull request https://github.com/zotero/translators/pull/915 forked from https://github.com/Juris-M/translators/blob/0e34b56e4f2efe9b5123addd4010f0e989ee36de/HeinOnline.js

### CODE REVIEW

File: HeinOnline.js [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+function getXPathStr(attr, elem, path) {
+	var res = ZU.xpath(elem, path);
+	res = res.length ? res[0][attr] : '';
+	return res ? res : '';
+}
+
+// Extract query values to keys on an object
+function extractQueryValues(url) {
+	var ret = {};
+	ret.base = url.replace(/(.*?)[a-zA-Z]+\?.*/, "$1");
```
Comment (adam3smith) [inline HeinOnline.js]: since URLs never has two question marks, how is this ever different from 
`url.replace(/[a-zA-Z]+\?.*/, "");` ?

File: HeinOnline.js [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+	for (var i=0,ilen=query.length;i<ilen;i++) {
+		var pair = query[i].split("=");
+		ret[pair[0]] = pair[1];
+	}
+	return ret;
+}
+
+// Remangle HeinOnline RIS to form expected by translator
+function fixRIS (ris) {
+	var lines = ris.split("\n");
```
Comment (adam3smith) [inline HeinOnline.js]: reposting aurimas' comment from the old review:
>You don't need to split up everything into single line. You can run the replacements on the whole text with mg flags (m makes ^ match start of line).

File: HeinOnline.js [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+		var pdfPageURL = doc.getElementsByClassName("updatediv");
+		ZU.doGet(risURL, function(ris) {
+			ris = fixRIS(ris);
+			if (pdfPageURL) {
+				ZU.doGet(pdfPageURL, function(pdfPage) {
+					// Call to pdfPageURL prepares PDF for download via META refresh URL
+					var pdfURL = null;
+					var m = pdfPage.match(/<META.*URL=([^"]+)/);
+					if (m) {
+						var pdfURL = m[1];
```
Comment (adam3smith) [inline HeinOnline.js]: from old review: no need for `var` here. The variable is declared in line 132

File: HeinOnline.js [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+	*********
+*/
+
+function detectWeb (doc, url) {
+	if (url.indexOf("/LuceneSearch?") > -1) {
+		if (getSearchResults(doc)) {
+			return "multiple";
+		}
+	} else {
+		return "journalArticle";
```
Comment (adam3smith) [inline HeinOnline.js]: from older review:
>Make sure either RIS out COinS is available

in other words, make sure that we only detect something when we're not then going to throw an error on import.

File: HeinOnline.js [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+
+			ZU.processDocuments(urls, scrapePage);
+		});
+	} else {
+		scrapePage(doc, url);
+	}
+}
+
+/** BEGIN TEST CASES **/
+var testCases = [
```
Comment (adam3smith) [inline HeinOnline.js]: as per aurimas, more tests cases would be great (if you're not using scaffold alredy, you'll definitely want to for that purpose, since it creates test JSON automatically.)]
https://www.zotero.org/support/dev/translators/scaffold

**Comment by adam3smith** [comment]
Great, almost there, just a couple of small things left. You can update your pull request by clicking the pencil icon here: https://github.com/zotero/translators/pull/1082/files and editing/pasting the style.

Any questions or problems, please let me know.

File: HeinOnline.js (hunk 1 above)
Comment (heinonline) [inline HeinOnline.js]: The first question mark is not a literal match, instead it is making the .* non-greedy.

Comment (adam3smith) [inline HeinOnline.js]: I get that, but when would that ever be necessary? Since there's only one ?, the non-greedy match has no effect. Maybe I'm just missing an example, but I'm thinking of a URL like
http://heinonline.org/HOL/Page?<somequerystring>  url.replace(/[a-zA-Z]+\?.*/, ""); chops of everything from Page? and returns the rest and your version keeps everything until HOL/ and returns it. In other words, the two are functionally equivalent, the 2nd one is less efficientl.

Comment (heinonline) [inline HeinOnline.js]: I see. Yes, you are correct.

File: HeinOnline.js (hunk 4 above)
Comment (heinonline) [inline HeinOnline.js]: I don't know how to do what you are asking, I'll look into it.

Comment (adam3smith) [inline HeinOnline.js]:
```
var COinS = getXPathStr("title", doc, '//span[contains(@class, "Z3988")]');
var RIS = getXPathStr("href", doc, '//form[@id="pagepicker"]//a[contains(@href, "PrintRequest")][1]');
...
else if (COinS || RIS) {
    return "journalArticle"
}
```

or something along those line.

File: HeinOnline.js [hunk 6]
```diff
@@ -2,14 +2,14 @@
 	"translatorID": "3dcbb947-f7e3-4bbd-a4e5-717f3701d624",
 	"label": "HeinOnline",
 	"creator": "Frank Bennett",
-	"target": "^https?://(www\\.)?heinonline\\.org/HOL/(?:LuceneSearch|Page|IFLPMetaData)\\?",
+	"target": "^https?://(www\.)?heinonline\.org/HOL/(?:LuceneSearch|Page|IFLPMetaData)\\?",
```
Comment (zuphilip) [inline HeinOnline.js]: Undo this change.

The point in a regexp refers to any character and therefore you have to escape it here `\.` in scaffold which will result in double escaped `\\.` here.

**Comment by heinonline** [comment]
Is there something else I need to do before this can be moved forward? Or is this just in a queue to be reviewed when time permits? I don't mean to be pushy I just want to make sure there is nothing I need to do. Thanks.

**Comment by zuphilip** [comment]
I think that there is just a queue for PR in general here. Maybe @adam3smith finds some time to check the update.

BTW there are no notifications if you update a PR and therefore it is a good idea to write a short message after you updated a PR.

**Comment by heinonline** [comment]
Thanks @zuphilip !

Pull request updated to address remaining issues please review.

**Comment by heinonline** [comment]
Just checking in to make sure there is nothing else to be done here. We really appreciate all your help with this.

**Comment by zuphilip** [comment]
I think that currently there is also vacation time:  :high_brightness: :palm_tree: :swimmer:

**Comment by heinonline** [comment]
We recently fixed the RIS file generator and so it no longer needs to be fixed here.

**Comment by adam3smith** [comment]
Great, thanks and sorry for the delay. FYI @fbennett your earlier work now fully implemented.

**Comment by fbennett** [comment]
Super, thanks for the follow-up.
